### PR TITLE
tree: basic set support for object-proxies

### PR DESCRIPTION
**Pull request by DLehenbauer**
Implements setting require/optional fields on object-like proxies

### CODE REVIEW

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts [hunk 1]
```diff
@@ -162,8 +171,30 @@ export function createObjectProxy<TSchema extends StructSchema, TTypes extends A
 				return undefined;
 			},
 			set(target, key, value) {
-				// TODO: Implement set
-				return false;
+				const fieldSchema = content.schema.structFields.get(key as FieldKey);
+
+				if (fieldSchema === undefined) {
+					// TODO: Verify that returning false will throw a TypeError in strict mode.
```
Comment (taylorsw04) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts:177]: The docs do claim this is true, but you're thinking it's worth a test?
Regardless, this made me wonder: will we ever run in sloppy mode? If so, should we just throw here?

**Review by taylorsw04 — APPROVED** [review]

**Review by taylorsw04 — APPROVED** [review]

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/object.spec.ts [hunk 2]
```diff
@@ -188,3 +192,153 @@ const tcs: TestCase[] = [
...
+			// 	{ list: before },
+			// 	(root) => {
+			// 		assert.deepEqual(root.list, before);
+			// 		root.list = after;
+			// 		assert.deepEqual(root.list, after);
+			// 	},
+			// );
+		});
+
+		describe.skip("optinal list", () => {
```
Comment (taylorsw04) [inline experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/object.spec.ts:312]:
```suggestion
		describe.skip("optional list", () => {
```

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts (hunk 1 above)
Comment (DLehenbauer) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts]: I added the missing test.  Thanks for the reminder.

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/lazyTree.ts [hunk 3]
```diff
@@ -509,6 +509,16 @@ const cachedStructClasses = new WeakMap<
...
 
+export function getBoxedField(
+	struct: LazyEntity,
+	key: FieldKey,
+	fieldSchema: FieldSchema,
+): TreeField {
+	return inCursorField(struct[cursorSymbol], key, (cursor) => {
+		return makeField(struct.context, fieldSchema, cursor);
+	});
+}
```
Comment (DLehenbauer) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/lazyTree.ts:520]: @CraigMacomber - I lifted this function as a "back door" to force an optional field into existence in order to set it.  LMK if you have a better alternative.

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts [hunk 4]
```diff
@@ -162,8 +171,29 @@ export function createObjectProxy<TSchema extends StructSchema, TTypes extends A
...
-				// TODO: Implement set
-				return false;
+				const fieldSchema = content.schema.structFields.get(key as FieldKey);
+
+				if (fieldSchema === undefined) {
+					return false;
+				}
+
+				// TODO: Is it safe to assume 'content' is a LazyEntity?
+				const field = getBoxedField(content as LazyEntity, key as FieldKey, fieldSchema);
```
Comment (DLehenbauer) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts:181]: @CraigMacomber - Ditto on better alternatives.

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/list.spec.ts [hunk 5]
```diff
@@ -23,15 +23,6 @@ const root = builder.struct("root", {
...
 describe("List", () => {
-	/** Similar to JSON stringify, but preserves 'undefined' and leaves numbers as-is. */
-	function pretty(arg: any) {
-		return arg === undefined
-			? "undefined"
-			: typeof arg === "number"
-			? arg
-			: JSON.stringify(arg);
-	}
-
```
Comment (DLehenbauer) [inline experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/list.spec.ts:34]: (Unrelated change to use shared copy from './utils')

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/object.spec.ts [hunk 6]
```diff
@@ -30,16 +32,18 @@ function testObjectLike(testCases: TestCase[]) {
...
-				});
+				itWithRoot(
+					`${pretty(real)} -> ${pretty(expected)}`,
+					schema,
+					initialTree,
+					(proxy) => {
+						const actual = fn(proxy as object);
+						assert.deepEqual(actual, expected);
+					},
+				);
```
Comment (DLehenbauer) [inline experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/object.spec.ts:46]: I lifted @noencke's 'itWithRoot' convenience to utils().
